pkg/bigquerysql/connector.go

```diff
@@ -0,0 +1,178 @@
+	incrementTableID string
+	storageURL       string
+
+	columns []cloudstorage.TableCol
+}
+
+func NewBigQueryConnector(bqClient *bigquery.Client, incrementTableID, datasetID, tableID string, storageURI *url.URL) (*BigQueryConnector, error) {
+	storageURL := fmt.Sprintf("%s://%s%s", storageURI.Scheme, storageURI.Host, storageURI.Path)
+	return &BigQueryConnector{
+		bqClient:         bqClient,
+		datasetID:        datasetID,
+		tableID:          tableID,
+		incrementTableID: incrementTableID,
+		storageURL:       storageURL,
+		columns:          nil,
+	}, nil
+}
+
+func (bc *BigQueryConnector) InitSchema(columns []cloudstorage.TableCol) error {
+	if len(bc.columns) != 0 {
+		return nil
+	}
+	if len(columns) == 0 {
+		return errors.New("Columns in schema is empty")
+	}
+	bc.columns = columns
+	log.Info("table columns initialized", zap.Any("Columns", columns))
+	return nil
+}
+
+func (bc *BigQueryConnector) ExecDDL(tableDef cloudstorage.TableDefinition) error {
+	if len(bc.columns) == 0 {
+		return errors.New("Columns not initialized. Maybe you execute a DDL before all DMLs, which is not supported now.")
+	}
+	ddls, err := GenDDLViaColumnsDiff(bc.datasetID, bc.tableID, bc.columns, tableDef)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(ddls) == 0 {
+		log.Info("No need to execute this DDL in Snowflake", zap.String("ddl", tableDef.Query))
+		return nil
+	}
+	// One DDL may be rewritten to multiple DDLs
+	ctx := context.Background()
+	for _, ddl := range ddls {
+		job, err := bc.bqClient.Query(ddl).Run(ctx)
+		if err != nil {
+			log.Error("Failed to execute DDL", zap.String("received", tableDef.Query), zap.String("rewritten", strings.Join(ddls, "\n")))
+			return errors.Annotate(err, fmt.Sprint("failed to execute", ddl))
+		}
+		status, err := job.Wait(ctx)
+		if err != nil {
+			log.Error("Failed to wait DDL", zap.String("received", tableDef.Query), zap.String("rewritten", strings.Join(ddls, "\n")))
+			return errors.Annotate(err, fmt.Sprint("failed to wait", ddl))
+		}
+		if status.Err() != nil {
+			log.Error("Failed to executed DDL", zap.String("received", tableDef.Query), zap.String("rewritten", strings.Join(ddls, "\n")))
+			return errors.Annotate(err, fmt.Sprint("failed to execute", ddl))
+		}
+	}
+	// update columns
+	bc.columns = tableDef.Columns
+	log.Info("Successfully executed DDL", zap.String("received", tableDef.Query), zap.String("rewritten", strings.Join(ddls, "\n")))
+	return nil
+}
+
+// CopyTableSchema copies table schema from TiDB to BigQuery
+// If table exists, delete it first
+func (bc *BigQueryConnector) CopyTableSchema(sourceDatabase string, sourceTable string, sourceTiDBConn *sql.DB) error {
+	ctx := context.Background()
+
+	tableExists, err := checkTableExists(ctx, bc.bqClient, bc.datasetID, bc.tableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if tableExists {
+		err = deleteTable(ctx, bc.bqClient, bc.datasetID, bc.tableID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
```

Comment:
https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#create_table_statement

Maybe we can use `create or replace` ?

Reply:
Good idea!

BigQuery provides two methods for interacting with the database:
- [API](https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#methods), which does not support create or replace, and have some bad design such as the use of `FLOAT`.
- SQL, as you mentioned.

I have now refactored all the DDL operations to use the `SQL` approach, which provides more comprehensive functionality. Please review again.

Reply:
👍